dnsapi/dns_wedos.sh

```diff
@@ -0,0 +1,366 @@
+
+# Function _wapi_post(), only takes data, prepares auth token and provide result
+# $1 - WAPI command string, like 'dns-domains-list'
+# $2 - WAPI data for given command, is not required
+# returns WAPI response if request were successfully delivered to WAPI endpoint
+_wapi_post() {
+  command=$1
+  data=$2
+
+  _debug "Command : ${command}"
+  _debug "Data : ${data}"
+
+  if [ -z "${command}" ]; then
+    _err "No command were provided, implamantation error!"
+    return 1
+  fi
+
+  # Prepare authentification token
+  hour=$(date +%H)
+  token=$(printf '%s' "${WEDOS_Authtoken}${hour}" | sha1sum | head -c 40)
+  _debug "Authentification token is '${token}'"
+
+  # Build xml request
+
+  request="request=<?xml version=\"1.0\" encoding=\"UTF-8\"?>\
+<request>\
+  <user>${WEDOS_Username}</user>\
+  <auth>${token}</auth>\
+  <command>${command}</command>"
+
+  if [ -z "${data}" ]; then
+    echo "" 1>/dev/null
+  else
+    request="${request}${data}"
+  fi
+
+  if [ -z "$TESTING_STAGE" ]; then
+    echo "" 1>/dev/null
+  else
+    request="${request}\
+  <test>1</test>"
+  fi
+
+  request="${request}\
+</request>"
+
+  _debug "Request to WAPI is: ${request}"
+
+  if ! response="$(_post "${request}" "$WEDOS_WAPI_ENDPOINT")"; then
+    _err "Error contacting WEDOS WAPI with command ${command}"
+    return 1
+  fi
+
+  _debug "Response : ${response}"
+  echo "${response}" | grep "<code>1000</code>" 1>/dev/null 2>/dev/null
+
+  return "$?"
+}
+
+# _get_root() function, for provided full domain, like _acme_challenge.www.example.com verify if WEDOS contains a primary active domain and found what is subdomain
+# $1 - full domain to verify, ie _acme_challenge.www.example.com
+# build ${_domain} found at WEDOS, like example.com and ${_sub_domain} from provided full domain, like _acme_challenge.www
+_get_root() {
+  domain=$1
+
+  if [ -z "${domain}" ]; then
+    _err "Function _get_root was called without argument, implementation error!"
+    return 1
+  fi
+
+  _debug "Get root for domain: ${domain}"
+
+  _debug "Getting list of domains using WAPI ..."
+
+  if ! _wapi_post "dns-domains-list"; then
+    _err "Error on WAPI request for list of domains, response : ${response}"
+    return 1
+  else
+    _debug "DNS list were successfully retrieved, response : ${response}"
+  fi
+
+  for xml_domain in $(echo "${response}" | tr -d '\012\015' | grep -o -E "<domain>( )*<name>.*</name>( )*<type>primary</type>( )*<status>active</status>" | grep -o -E "<name>.*</name>"); do
+    _debug "Active and primary XML DOMAIN found: ${xml_domain}"
+    end_of_name=$((${#xml_domain} - 7))
+    xml_domain_name=$(echo "${xml_domain}" | cut -c 7-${end_of_name})
+    _debug "Found primary active domain: ${xml_domain_name}"
+    regex=".*\\."$(echo "${xml_domain_name}" | sed 's/\./\\./g')
+    _debug "Regex for matching domain: '${regex}'"
+
+    if ! echo "${domain}" | grep -E "${regex}" 1>/dev/null 2>/dev/null; then
```

Comment:
use `_egrep_o`

Reply:
All reimplemented using sed and tr only.